[pytorch] fix ParallelNative.h/cpp build

ljk53 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26702 [pytorch][perf] add mobile friendly at:parallel_for backend
* **#26746 [pytorch] fix ParallelNative.h/cpp build**
* #26783 [pytorch][mobile] remove backward functions from jit-op-registry for mobile build

Summary:
PR #25439 introduced type() to IValue, which enforced ivalue::Future to
be constructed with type. This broke ParallelNative.h and a few other
places where Future was constructed without type and simply used as a barrier:
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/ParallelNative.h#L63

Use NoneType::get() to initialize these Futures.

Test Plan:
- builds

Differential Revision: [D17556002](https://our.internmc.facebook.com/intern/diff/D17556002)

ljk53 committed:
[pytorch] add default constructor to ivalue::Future

Summary:
PR #25439 introduced type() to IValue, which enforced ivalue::Future to
be constructed with type. This broke ParallelNative.h and a few other
places where Future was constructed without type and simply used as a barrier:
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/ParallelNative.h#L63

We could set NoneType by default if we want to enforce non-null type
everywhere, but looks like Tuple can take null TupleType so just follow
it.

Test Plan:
- builds

[ghstack-poisoned]

smessmer commented on aten/src/ATen/core/ivalue_inl.h:
```diff
@@ -191,6 +191,7 @@ struct C10_EXPORT ivalue::Future final : c10::intrusive_ptr_target {
   }
 
  public:
+  Future() {}
```
nit: `Future() = default;` is better

smessmer reviewed (commented):
Are we sure we want to be able to construct Futures without type information? Afaik, we want to have types everywhere. cc @zdevito

ljk53 committed:
Update on "[pytorch] add default constructor to ivalue::Future"


Summary:
PR #25439 introduced type() to IValue, which enforced ivalue::Future to
be constructed with type. This broke ParallelNative.h and a few other
places where Future was constructed without type and simply used as a barrier:
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/ParallelNative.h#L63

We could set NoneType by default if we want to enforce non-null type
everywhere, but looks like Tuple can take null TupleType so just follow
it.

Test Plan:
- builds

Differential Revision: [D17556002](https://our.internmc.facebook.com/intern/diff/D17556002)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch] add default constructor to ivalue::Future"


Summary:
PR #25439 introduced type() to IValue, which enforced ivalue::Future to
be constructed with type. This broke ParallelNative.h and a few other
places where Future was constructed without type and simply used as a barrier:
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/ParallelNative.h#L63

We could set NoneType by default if we want to enforce non-null type
everywhere, but looks like Tuple can take null TupleType so just follow
it.

Test Plan:
- builds

Differential Revision: [D17556002](https://our.internmc.facebook.com/intern/diff/D17556002)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch] add default constructor to ivalue::Future"


Summary:
PR #25439 introduced type() to IValue, which enforced ivalue::Future to
be constructed with type. This broke ParallelNative.h and a few other
places where Future was constructed without type and simply used as a barrier:
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/ParallelNative.h#L63

We could set NoneType by default if we want to enforce non-null type
everywhere, but looks like Tuple can take null TupleType so just follow
it.

Test Plan:
- builds

Differential Revision: [D17556002](https://our.internmc.facebook.com/intern/diff/D17556002)

[ghstack-poisoned]

suo reviewed (changes requested):
I don't think this makes a ton of sense. TupleType takes a null type because the type can be lazily inferred by inspecting the types of the contained elements, not because it actually doesn't have a type.

I would prefer not to default construct `Future` as there is not an obvious default. Can we change ParallelNative and friends to use `Future<None>`?

ilia-cher commented:
@suo I wonder would it make sense to make it None by default (with default ctor)

ljk53 committed:
Update on "[pytorch] fix ParallelNative.h/cpp build"


Summary:
PR #25439 introduced type() to IValue, which enforced ivalue::Future to
be constructed with type. This broke ParallelNative.h and a few other
places where Future was constructed without type and simply used as a barrier:
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/ParallelNative.h#L63

Use NoneType::get() to initialize these Futures.

Test Plan:
- builds

Differential Revision: [D17556002](https://our.internmc.facebook.com/intern/diff/D17556002)

[ghstack-poisoned]

ljk53 commented on aten/src/ATen/ParallelNative.h:
```diff
@@ -60,7 +60,10 @@ inline void parallel_for(
+    std::vector<std::shared_ptr<c10::ivalue::Future>> futures(num_tasks);
+    for (size_t i = 0; i < num_tasks; ++i) {
+      futures[i] = std::make_shared<c10::ivalue::Future>(NoneType::get());
+    }
```
@ilia-cher let me know if I should move this to a common util function, maybe in Parallel.h? I can reuse it in ParallelNativeMobile.h as well.

zdevito commented:
This and https://github.com/pytorch/pytorch/pull/26739 are duplicates, fyi.

ljk53 committed:
Update on "[pytorch] fix ParallelNative.h/cpp build"


Summary:
PR #25439 introduced type() to IValue, which enforced ivalue::Future to
be constructed with type. This broke ParallelNative.h and a few other
places where Future was constructed without type and simply used as a barrier:
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/ParallelNative.h#L63

Use NoneType::get() to initialize these Futures.

Test Plan:
- builds

Differential Revision: [D17556002](https://our.internmc.facebook.com/intern/diff/D17556002)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch] fix ParallelNative.h/cpp build"


Summary:
PR #25439 introduced type() to IValue, which enforced ivalue::Future to
be constructed with type. This broke ParallelNative.h and a few other
places where Future was constructed without type and simply used as a barrier:
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/ParallelNative.h#L63

Use NoneType::get() to initialize these Futures.

Test Plan:
- builds

Differential Revision: [D17556002](https://our.internmc.facebook.com/intern/diff/D17556002)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch] fix ParallelNative.h/cpp build"


Summary:
PR #25439 introduced type() to IValue, which enforced ivalue::Future to
be constructed with type. This broke ParallelNative.h and a few other
places where Future was constructed without type and simply used as a barrier:
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/ParallelNative.h#L63

Use NoneType::get() to initialize these Futures.

Test Plan:
- builds

Differential Revision: [D17556002](https://our.internmc.facebook.com/intern/diff/D17556002)

[ghstack-poisoned]
